arch/arm/core/aarch64/smp.c

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2020 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+/**
+ * @file
+ * @brief codes required for AArch64 multicore and Zephyr smp support
+ */
+
+#include <cache.h>
+#include <device.h>
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <ksched.h>
+#include <soc.h>
+#include <init.h>
+#include <arch/arm/aarch64/arm_mmu.h>
+#include <arch/arm/aarch64/barrier.h>
+#include <arch/cpu.h>
+#include <drivers/interrupt_controller/gic.h>
+#include <drivers/psci.h>
+#include <sys/arch_interface.h>
+
+#define SGI_SCHED_IPI	0
+
+volatile struct {
+	void *sp; /* Fixed at the first entry */
+	arch_cpustart_t fn;
+	void *arg;
+	char pad[] __aligned(L1_CACHE_BYTES);
+} arm64_cpu_init[CONFIG_MP_NUM_CPUS];
+
+/*
+ * _curr_cpu is used to record the struct of _cpu_t of each cpu.
+ * for efficient usage in assembly
+ */
+volatile _cpu_t *_curr_cpu[CONFIG_MP_NUM_CPUS];
+
+extern void __start(void);
+/* Called from Zephyr initialization */
+void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
+		    arch_cpustart_t fn, void *arg)
+{
+	const struct device *psci;
+
+	__ASSERT(sizeof(arm64_cpu_init[0]) == ARM64_CPU_INIT_SIZE,
+		 "ARM64_CPU_INIT_SIZE != sizeof(arm64_cpu_init[0]\n");
+
+	_curr_cpu[cpu_num] = &(_kernel.cpus[cpu_num]);
+	arm64_cpu_init[cpu_num].fn = fn;
+	arm64_cpu_init[cpu_num].arg = arg;
+	arm64_cpu_init[cpu_num].sp =
+		(void *)(Z_THREAD_STACK_BUFFER(stack) + sz);
+
+	arch_dcache_range((void *)&arm64_cpu_init[cpu_num],
+			  sizeof(arm64_cpu_init[cpu_num]), K_CACHE_WB_INVD);
+
+	psci = device_get_binding("PSCI");
+	/* TODO: get mpidr from device tree, using cpu_num */
+	if (psci_cpu_on(psci, cpu_num, (uint64_t)&__start))
+		printk("Failed to boot CPU%d\n", cpu_num);
+
+	/* Wait secondary cores up, see z_arm64_secondary_start */
+	while (arm64_cpu_init[cpu_num].fn) {
+		wfe();
+	}
+}
+
+/* the C entry of secondary cores */
+void z_arm64_secondary_start(void)
+{
+	arch_cpustart_t fn;
+	int cpu_num = MPIDR_TO_CORE(GET_MPIDR());
+
+	z_arm64_mmu_init();
+
+#ifdef CONFIG_SMP
+	arm_gic_secondary_init();
+
+	irq_enable(0);
+#endif
+
+	fn = arm64_cpu_init[cpu_num].fn;
+
```

Comment:
secondary core clears .fn  to announce its presence.  Primary core is polling for this.